util/conversion/conversion.go

```diff
@@ -138,28 +137,66 @@ func GetFuzzer(scheme *runtime.Scheme, funcs ...fuzzer.FuzzerFuncs) *fuzz.Fuzzer
 
 // FuzzTestFunc returns a new testing function to be used in tests to make sure conversions between
 // the Hub version of an object and an older version aren't lossy.
-func FuzzTestFunc(scheme *runtime.Scheme, hub conversion.Hub, dst conversion.Convertible, funcs ...fuzzer.FuzzerFuncs) func(*testing.T) {
+func FuzzTestFunc(scheme *runtime.Scheme, hub conversion.Hub, dst conversion.Convertible, beforeCompare func(c1, c2 conversion.Convertible),
+	funcs ...fuzzer.FuzzerFuncs) func(*testing.T) {
 	return func(t *testing.T) {
-		g := gomega.NewWithT(t)
-		fuzzer := GetFuzzer(scheme, funcs...)
-
-		for i := 0; i < 10000; i++ {
-			// Make copies of both objects, to avoid changing or re-using the ones passed in.
-			hubCopy := hub.DeepCopyObject().(conversion.Hub)
-			dstCopy := dst.DeepCopyObject().(conversion.Convertible)
-
-			// Run the fuzzer on the Hub version copy.
-			fuzzer.Fuzz(hubCopy)
-
-			// Use the hub to convert into the convertible object.
-			g.Expect(dstCopy.ConvertFrom(hubCopy)).To(gomega.Succeed())
-
-			// Make another copy of hub and convert the convertible object back to the hub version.
-			after := hub.DeepCopyObject().(conversion.Hub)
-			g.Expect(dstCopy.ConvertTo(after)).To(gomega.Succeed())
-
-			// Make sure that the hub before the conversions and after are the same, include a diff if not.
-			g.Expect(apiequality.Semantic.DeepEqual(hubCopy, after)).To(gomega.BeTrue(), cmp.Diff(hubCopy, after))
-		}
+		t.Run("spoke-hub-spoke", func(t *testing.T) {
+			g := gomega.NewWithT(t)
+			fuzzer := GetFuzzer(scheme, funcs...)
+
+			for i := 0; i < 10000; i++ {
+				// Create hub object
+				hubExisting := hub.DeepCopyObject().(conversion.Hub)
+				fuzzer.Fuzz(hubExisting)
+
+				// Convert hub object to spoke
+				spokeFirstGet := dst.DeepCopyObject().(conversion.Convertible)
+				g.Expect(spokeFirstGet.ConvertFrom(hubExisting)).To(gomega.Succeed())
+
+				// Do changes in the spoke
+				fuzzer.Fuzz(spokeFirstGet)
+				// Fuzz() might delete the annotation containing the hub serialized. So, re-add it.
+				MarshalData(hubExisting.(metav1.Object), spokeFirstGet.(metav1.Object))
+
+				// Convert the changed spoke to hub
+				hubUpdated := hub.DeepCopyObject().(conversion.Hub)
+				g.Expect(spokeFirstGet.ConvertTo(hubUpdated)).To(gomega.Succeed())
+
+				// Convert hub back to spoke and check if the changed spoke is still the same after spoke --> hub --> spoke conversion
+				spokeSecondGet := dst.DeepCopyObject().(conversion.Convertible)
+				g.Expect(spokeSecondGet.ConvertFrom(hubUpdated)).To(gomega.Succeed())
+
+				// Override any fields in spokes before checking for equality
+				// This is typically used for copying over deprecated fields that are not required for comparison
+				if beforeCompare != nil {
+					beforeCompare(spokeFirstGet, spokeSecondGet)
+				}
+
+				g.Expect(apiequality.Semantic.DeepEqual(spokeFirstGet, spokeSecondGet)).To(gomega.BeTrue(), cmp.Diff(spokeFirstGet, spokeSecondGet))
```

Comment:
Could we avoid this somehow? 

If we're dropping a field, and we expect the data to be lost, maybe the comparison here shouldn't be a DeepEqual. I'm not 100% sure if we want to lose the data either, one alternative would be to store the dropped fields in an annotation, but that would be stored in etcd, which we probably don't want to do.

@ncdc @CecileRobertMichon thoughts?

Reply:
For alternatives to DeepEqual, I looked into using `Equalities.AddFunc`, and bypass deprecated fields to avoid copying over data. The problem was that we can only do it for a type (`ObjectMeta`), and there are other places where this type is used, which may result in unintended consequences.

Reply:
I think these are all the options I can come up with:

1. Don't ever drop fields. I don't think this is a viable option.

1. Don't do spoke-hub-spoke tests. The lack of these resulted in broken conversion functions that were only discovered in manual testing. I don't think this is a viable option.

1. Store the dropped fields somewhere (annotation, presumably) so we don't lose them. As you pointed out @vincepri, they would be stored in etcd, which is not desirable. Also, if we have dropped a field in a newer API version, an older client attempting to get or set this field in the older API version presumably won't really be affecting anything of value. Any controller talking the older API version likely isn't using the field any more (if it is, then I'd wager we're converting the field and are not dropping it). Any controller talking the newer API version definitely isn't doing anything with the now-nonexistent field.

1. Allow the unit test author to define ways to manipulate `spokeFirstGet` and `spokeSecondGet`, as is done in this PR.

1. Use custom Equalities functions. Even though this is type-based, I think we could add a custom function for MachineSpec maybe, given that doing it with ObjectMeta doesn't get us what we need.

Reply:
I thought of one more, which may not require any signature change:

6. Pass in a custom `FuzzerFuncs` (the optional last varargs argument in the function) that fuzzes a v1alpha2 MachineSpec, making sure ObjectMeta is never fuzzed with random data. Try that and see if it works?

Reply:
> 5. Use custom Equalities functions. Even though this is type-based, I think we could add a custom function for MachineSpec maybe, given that doing it with ObjectMeta doesn't get us what we need.

This is a possibility although the function would be a bit complex, since we'll need to call `DeepEqual` on the rest of the `MachineSpec` fields inside this function.

> 6. Pass in a custom FuzzerFuncs (the optional last varargs argument in the function) that fuzzes a v1alpha2 MachineSpec, making sure ObjectMeta is never fuzzed with random data. Try that and see if it works?

I think I tried this and gave up for some reason that I don't remember :) Will give another shot at this, though. This would also avoid any breaking changes.

Reply:
Haven't had time to think about this enough, if we can find a way to detect which fields have been dropped, we could remove wipe out those fields from both objects, instead of retaining the data.

Reply:
@ncdc @vincepri PTAL at [this](https://github.com/shysank/cluster-api/pull/1), and let me know if it aligns with what you were thinking.

Reply:
How about disabling this test for now, given that's blocking other PRs as well and open an issue to revisit and fix it later?

Reply:
👍 to disabling for now. This needs more thought. (Sorry, I didn't think this would be such a thorny issue!)

Reply:
created https://github.com/kubernetes-sigs/cluster-api/issues/3923 to skip spoke-hub-spoke tests.